### PR TITLE
Escape latex before markdown rendering in PDF generation

### PR DESCRIPTION
**Pull request by nop33**

### CODE REVIEW

File: indico/MaKaC/PDFinterface/base.py [hunk 1]
```diff
@@ -708,8 +709,21 @@ def __init__(self):
...
+        # Escape LaTeX before markdown
+        def _convert_markdown(text):
+            math_segments = []
+            placeholder = u"\uE000"
+
+            def _math_replace(m):
+                math_segments.append(m.group(0))
+                return placeholder
+
+            text = re.sub(r'\$[^\$]+\$|\$\$(^\$)\$\$', _math_replace, text)
```
Comment (ThiefMaster) [inline indico/MaKaC/PDFinterface/base.py]: `latex_escape` in mdx_latex.py seems to do almost the same thing - why do we have to do it here manually? Bug in mdx_latex?

Comment (nop33) [inline indico/MaKaC/PDFinterface/base.py]: Looks like the `LaTeXExtension.extendMarkdown` method is not escaping latex. However I totally missed the `latex_escape` function. I will see if using it fixes the problem. Thanks for the observation.

Comment (nop33) [inline indico/MaKaC/PDFinterface/base.py]: `latex_escape` won't convert the Markdown.

Comment (pferreir) [inline indico/MaKaC/PDFinterface/base.py]: This is a situation where a Ruby block would have rocked :/

**Comment by pferreir** [comment]
As you can see, the tests break :wink:

File: indico/util/string.py [hunk 2]
```diff
@@ -190,18 +190,23 @@ def remove_tags(text):
...
+        placeholder = u"\uE000"
 
-    def math_replace(m):
-        math_segments.append(m.group(0))
-        return placeholder
+        def _math_replace(m):
+            math_segments.append(m.group(0))
+            return placeholder
+
+        text = re.sub(r'\$[^\$]+\$|\$\$(^\$)\$\$', _math_replace, to_unicode(text))
```
Comment (pferreir) [inline indico/util/string.py]: OTOH, this looks exactly the same as https://github.com/indico/indico/pull/1881/files#diff-b0098f64f333656ff7eb7ffbc29af757R714

Comment (nop33) [inline indico/util/string.py]: What do you think of this approach?
https://gist.github.com/nop33/0201a6b952a1f8a6dd4a

Comment (nop33) [inline indico/util/string.py]: ping!

Comment (pferreir) [inline indico/util/string.py]: A bit confusing. It would probably be better to pass a callback.

Comment (nop33) [inline indico/util/string.py]: Do you mean something like this?

```python
# MaKaC/PDFinterface/base.py
 
class PDFLaTeXBase(object):
 
    _table_of_contents = False
 
    def __init__(self):
 
        # Markdown -> LaTeX renderer
        # safe_mode - strip out all HTML
        md = markdown.Markdown(safe_mode='remove')
        latex_mdx = mdx_latex.LaTeXExtension()
        latex_mdx.extendMarkdown(md, markdown.__dict__)
 
        def _convert_markdown(text):
            return render_markdown(text, md_convert=md.convert)
 
        self._args = {
            'md_convert': _convert_markdown
        }
    ...
    
# indico/util/string.py
def render_markdown(text, escape_latex_math=True, md_convert=None, **kwargs):
    """ Mako markdown to html filter """
    if escape_latex_math:
        math_segments = []
        placeholder = u"\uE000"
 
        def _math_replace(m):
            math_segments.append(m.group(0))
            return placeholder
 
        text = re.sub(r'\$[^\$]+\$|\$\$(^\$)\$\$', _math_replace, to_unicode(text))
 
    if md is not None:
        return re.sub(placeholder, lambda _: math_segments.pop(0), md_convert(text))
    res = markdown.markdown(bleach.clean(text, tags=BLEACH_ALLOWED_TAGS), **kwargs)
 
    if escape_latex_math:
        res = re.sub(placeholder, lambda _: math_segments.pop(0), res)
    return res.encode('utf-8')
```

Comment (pferreir) [inline indico/util/string.py]: I guess you mean `if md_convert is not None`.
But I'm not convinced... I guess it's better to discuss this IRL.

**Comment by pferreir** [comment]
Merged in 47b4b9127ca726b670e90dbb46ab15f828e96086.

**Pull request closed without merging**